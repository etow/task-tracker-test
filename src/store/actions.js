import taskRepository from '../repositories/taskRepository';
import taskCategoryRepository from '../repositories/taskCategoryRepository';
import {
  SET_TASKS_CATEGORIES,
  SET_TASKS_CATEGORIES_DATA,
  SET_TASKS,
  ADD_TASK,
  DELETE_TASK,
  SET_TASK_TO_EDIT,
  SET_TASK,
  FETCH_TASKS_CATEGORIES,
  FETCH_TASKS,
  CREATE_TASK,
  UPDATE_TASK,
  ROLLBACK_STATE,
  BACKUP_STATE,
  UPDATE_TASKS,
} from './constants';

export default {
  [FETCH_TASKS_CATEGORIES]: ({ commit }) => {
    return taskCategoryRepository.get()
      .then((response) => {
        commit(SET_TASKS_CATEGORIES, response.data);
        commit(SET_TASKS_CATEGORIES_DATA, response.data);
      });
  },
  [FETCH_TASKS]: ({ commit }) => {
    return taskRepository.get()
      .then((response) => commit(SET_TASKS, response.data));
  },
  [CREATE_TASK]: ({ commit, state }, newTask) => {
    commit(BACKUP_STATE, state);
    const task = {
      ...newTask,
      id: Math.random().toString(36).substr(2, 6),
      order: state.tasks[newTask.category].length,
    };
    commit(ADD_TASK, task);
    return taskRepository.create(task);
  },
  [DELETE_TASK]: ({ commit, state }, {task, category}) => {
    commit(BACKUP_STATE, state);
    commit(DELETE_TASK, {task, category});
    commit(SET_TASK_TO_EDIT, {});
    return taskRepository.delete(task.id);
  },
  [UPDATE_TASK]: ({ commit, state }, {task, prevCategory}) => {
    commit(BACKUP_STATE, state);
    if (prevCategory && task.category !== prevCategory) {
      commit(DELETE_TASK, { task, category: prevCategory });
    }
    const targetIndex = state.tasks[task.category].findIndex((aTask) => aTask.id === task.id);
    if (targetIndex >= 0) {
      commit(SET_TASK, { task, targetIndex });
    } else {
      const order = state.tasks[task.category].length;
      commit(ADD_TASK, { ...task, order });
    }
    commit(SET_TASK_TO_EDIT, {})
    return taskRepository.update(task);
  },
  [UPDATE_TASKS]: ({ commit, state }, { tasks, category }) => {
    commit(BACKUP_STATE, state);
    tasks.forEach((task, index) => {
      task.category = category;
      task.order = index;
    });
    commit(UPDATE_TASKS, { tasks, category });

    return taskRepository.update(tasks);
  },
  [SET_TASK_TO_EDIT]: ({ commit, state}, task) => {
    commit(BACKUP_STATE, state);
    commit(SET_TASK_TO_EDIT, task);
  },
  [ROLLBACK_STATE]: ({ commit }) => {
    commit(ROLLBACK_STATE);
  },
};
