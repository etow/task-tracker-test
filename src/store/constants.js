

/* mutations and actions */
export const ADD_TASK = 'ADD_TASK';
export const SET_TASKS = 'SET_TASKS';
export const SET_TASKS_CATEGORIES = 'SET_TASKS_CATEGORIES';
export const SET_TASKS_CATEGORIES_DATA = 'SET_TASKS_CATEGORIES_DATA';
export const SET_TASK = 'SET_TASK';
export const SET_TASK_TO_EDIT = 'SET_TASK_TO_EDIT';
export const DELETE_TASK = 'DELETE_TASK';
export const FETCH_TASKS_CATEGORIES = 'FETCH_TASKS_CATEGORIES';
export const FETCH_TASKS = 'FETCH_TASKS';
export const CREATE_TASK = 'CREATE_TASK';
export const UPDATE_TASK = 'UPDATE_TASK';
export const ROLLBACK_STATE = 'ROLLBACK_STATE';
export const BACKUP_STATE = 'BACKUP_STATE';
export const UPDATE_TASKS = 'UPDATE_TASKS';
export const UPDATE_TASK_ACTIVITY = 'UPDATE_TASK_ACTIVITY';
