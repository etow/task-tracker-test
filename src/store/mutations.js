import {
  ADD_TASK,
  SET_TASKS,
  SET_TASKS_CATEGORIES,
  SET_TASKS_CATEGORIES_DATA,
  SET_TASK,
  SET_TASK_TO_EDIT,
  DELETE_TASK,
  ROLLBACK_STATE,
  BACKUP_STATE,
  UPDATE_TASKS,
  UPDATE_TASK_ACTIVITY
} from './constants';

export default {
  [ADD_TASK]: (state, task) => {
    state.tasks[task.category].push(task);
  },
  [SET_TASKS]: (state, tasks) => {
    Object.keys(state.tasks)
      .forEach((category) => {
        const tasksByCategory = tasks.filter((task) => task.category === category);
        state.tasks[category] = tasksByCategory.sort((a, b) => a.order - b.order);
      });
  },
  [SET_TASKS_CATEGORIES]: (state, categories) => {
    categories.forEach((cat) => state.tasks[cat.name] = []);
  },
  [SET_TASKS_CATEGORIES_DATA]: (state, categories) => {
    categories.forEach((cat) => state.categories[cat.name] = cat);
  },
  [SET_TASK]: (state, { task, targetIndex }) => {
    const index = targetIndex || state.tasks[task.category].findIndex((aTask) => aTask.id === task.id);
    state.tasks[task.category][index] = task;
  },
  [SET_TASK_TO_EDIT]: (state, task) => {
    state.taskToEdit = task;
  },
  [DELETE_TASK]: (state, {task, category}) => {
    state.tasks[category] = state.tasks[category].filter((aTask) => aTask.id !== task.id);
  },
  [ROLLBACK_STATE]: (state) => {
    Object.assign(state, JSON.parse(JSON.stringify(state.prevState)));
  },
  [BACKUP_STATE]: (state, currentState) => {
    state.prevState = JSON.parse(JSON.stringify(currentState));
  },
  [UPDATE_TASKS]: (state, {tasks, category}) => {
    state.tasks[category] = tasks;
  },
  [UPDATE_TASK_ACTIVITY]: (state, tasks) => {
    if (!Array.isArray(tasks)) {
      tasks = [tasks];
    }
    tasks.forEach((task) => {
      const found = state.tasks[task.category].find((aTask) => aTask.id === task.id);
      found.activity = task.activity;
    });
  }
};
