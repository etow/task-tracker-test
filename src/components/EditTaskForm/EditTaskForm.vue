<template>
  <div>
    <ElForm
      :model="form"
      class="edit-task-form"
      ref="form"
      :rules="formValidation"
      label-width="120px"
      label-position="left">
      <ElFormItem label="Name" prop="name">
        <ElInput v-model="form.name" />
      </ElFormItem>
      <ElFormItem label="Description" prop="description">
        <ElInput v-model="form.description" type="textarea" />
      </ElFormItem>
      <ElFormItem label="Status" prop="category">
        <ElSelect v-model="form.category" style="width: 100%">
          <ElOption v-for="(category) in categories" :key="category.name" :value="category.name" />
        </ElSelect>
      </ElFormItem>
      <ElFormItem label="Estimate" prop="estimate">
        <ElInputNumber v-model="form.estimate" controls-position="right" :min="0" />
      </ElFormItem>
      <ElFormItem style="margin-left: -120px; padding-top: 20px;">
        <div style="display: flex;width: 100%;">
          <ElPopconfirm
            title="Are you sure to delete this task?"
            hide-icon="true"
            confirm-button-type="danger"
            @confirm="handleDelete">
            <template #reference>
              <ElButton type="text" style="color: #f56c6c;">Delete</ElButton>
            </template>
          </ElPopconfirm>
          <div style="margin-left: auto;">
            <ElButton @click="handleCancel">Cancel</ElButton>
            <ElButton type="primary" @click="handleSave">Save</ElButton>
          </div>
        </div>
      </ElFormItem>
    </ElForm>
  </div>
</template>

<script>
import {
  ElMessage,
  ElForm,
  ElFormItem,
  ElInput,
  ElSelect,
  ElOption,
  ElPopconfirm,
  ElButton,
  ElInputNumber,
  ElLoading,
} from 'element-plus';
import { UPDATE_TASK, DELETE_TASK, SET_TASK_TO_EDIT, ROLLBACK_STATE } from '../../store/constants';

export default {
  data() {
    return {
      form: {
        name: '',
        description: '',
        category: '',
        estimate: 1,
        activity: {},
      },
      formValidation: {
        name: [
          {
            required: true,
            message: 'Name is required',
            trigger: 'blur',
          },
        ],
      },
      loading: false,
    };
  },
  components: {
    ElForm,
    ElFormItem,
    ElInput,
    ElSelect,
    ElOption,
    ElPopconfirm,
    ElButton,
    ElInputNumber,
  },
  watch: {
    taskToEdit: {
      handler(task) {
        this.form = { ...task};
      },
    },
  },
  mounted() {
    this.form = { ...this.taskToEdit};
  },
  computed: {
    categories() {
      return this.$store.state.categories;
    },
    taskToEdit() {
      return this.$store.state.taskToEdit;
    },
  },
  methods: {
    handleSave() {
      this.$refs.form.validate((valid) => {
        if (valid) {
          this.loading = ElLoading.service({ target: 'form' });
          this.$store.dispatch(UPDATE_TASK, { task: this.form, prevCategory: this.taskToEdit.category })
          .catch((err) => {
            console.error(err.message);
            ElMessage({
              message: 'Something went wrong while trying to save the changes, please try again',
              type: 'error',
              duration: 5000,
            });
            this.$store.dispatch(ROLLBACK_STATE);
          })
          .finally(() => {
            this.loading.close();
            this.$store.dispatch(SET_TASK_TO_EDIT, {});
            this.$emit('save');
          })
        }
      });
    },
    handleDelete() {
      this.loading = ElLoading.service({ target: 'form' });
      this.$store.dispatch(DELETE_TASK, { task: this.form, category: this.taskToEdit.category })
      .catch((err) => {
        console.error(err.message);
        ElMessage({
          message: 'Something went wrong while trying to delete the task, please try again',
          type: 'error',
          duration: 5000,
        });
        this.$store.dispatch(ROLLBACK_STATE);
      })
      .finally(() => {
        this.$store.dispatch(SET_TASK_TO_EDIT, {});
        this.$emit('delete');
        this.loading.close();
      });
    },
    handleCancel() {
      this.$store.dispatch(SET_TASK_TO_EDIT, {});
      this.$emit('cancel');
    },
  },
};
</script>
<style scoped>
.edit-task-form {
  position: relative;
}
</style>
