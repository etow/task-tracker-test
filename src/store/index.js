import { createStore } from 'vuex';

import mutations from './mutations';
import actions from './actions';

export default createStore({
  state: {
    tasks: {},
    categories: {},
    taskToEdit: {},
    prevState: {},
  },
  mutations,
  actions,
});
