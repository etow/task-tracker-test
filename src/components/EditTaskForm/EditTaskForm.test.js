import { shallowMount, flushPromises, config } from '@vue/test-utils';
import { ElMessage } from 'element-plus';
import { createStore } from 'vuex';
import EditTaskForm from './EditTaskForm.vue';

import { UPDATE_TASK, DELETE_TASK, SET_TASK_TO_EDIT, ROLLBACK_STATE } from '../../store/constants';


jest.mock('element-plus', () => ({
  ElMessage: jest.fn(),
  ElLoading: { service: jest.fn(() => ({ close: jest.fn() })) },
}));

const tasks = [
  { name: 'test1', category: 'Planned', order: 0, estimate: 0, activity: {} },
  { name: 'test2', category: 'Planned', order: 1, estimate: 0, activity: {} },
  { name: 'test3', category: 'Planned', order: 2, estimate: 0, activity: {} },
];

beforeAll(() => {
  config.renderStubDefaultSlot = true;
});

afterAll(() => {
  config.renderStubDefaultSlot = false;
});

describe('EditTaskForm.vue', () => {
  it('renders correctly', async () => {
    // mock store
    const [store] = mockStore();
    const wrapper = createWrapper((dataToMount) => {
      dataToMount.global.plugins = [store];
      return dataToMount;
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  test('form pass validation and save the changes', async () => {
    // mock store
    const formValidation = jest.fn(cb => cb(true));

    const [store, { actions }] = mockStore();
    const wrapper = createWrapper((dataToMount) => {
      dataToMount.global.plugins = [store];
      dataToMount.global.stubs.ElForm = { template: '<form><slot /></form>', methods: { validate: formValidation }};
      return dataToMount;
    });

    wrapper.vm.form = { name: 'Test Save', category: 'Planned', order: 1000 };
    wrapper.vm.handleSave();
    expect(formValidation).toHaveBeenCalled();
    await wrapper.vm.$nextTick()
    expect(actions[UPDATE_TASK]).toHaveBeenCalledWith(expect.anything(), { task: wrapper.vm.form, prevCategory: store.state.taskToEdit.category });
  });

  test('form fails validation and save is not tiggered', async () => {
    const formValidation = jest.fn(cb => cb(false));

    const [store, { actions }] = mockStore();
    const wrapper = createWrapper((dataToMount) => {
      dataToMount.global.plugins = [store];
      dataToMount.global.stubs.ElForm = { template: '<form><slot /></form>', methods: { validate: formValidation }};
      return dataToMount;
    });

    wrapper.vm.handleSave();
    expect(formValidation).toHaveBeenCalled();
    await wrapper.vm.$nextTick()
    expect(actions[UPDATE_TASK]).toHaveBeenCalledTimes(0);
  });

  test('handle delete', async () => {
    const [store, { actions }] = mockStore({ taskToEdit: tasks[0]});
    const wrapper = createWrapper((dataToMount) => {
      dataToMount.global.plugins = [store];
      return dataToMount;
    });
    wrapper.vm.handleDelete();
    await wrapper.vm.$nextTick()
    expect(actions[DELETE_TASK]).toHaveBeenCalledWith(expect.anything(), { task: wrapper.vm.form, category: tasks[0].category });
    await flushPromises();
  });

  test('handle cancel', async () => {
    const [store] = mockStore();
    const wrapper = createWrapper((dataToMount) => {
      dataToMount.global.plugins = [store];
      return dataToMount;
    });

    wrapper.vm.handleCancel();
    await wrapper.vm.$nextTick()
  });

  it('it handles error when updating a task', async () => {
    jest.spyOn(console, 'error').mockImplementation(() => {});
    const formValidation = jest.fn(cb => cb(true));
    const [store, { actions }] = mockStore({ update_task: jest.fn(() => Promise.reject('error'))});
    const wrapper = createWrapper((dataToMount) => {
      dataToMount.global.plugins = [store];
      dataToMount.global.stubs.ElForm = { template: '<form><slot /></form>', methods: { validate: formValidation }};
      return dataToMount;
    });

    wrapper.vm.form = { name: 'Test Save', category: 'Planned', order: 1000 };
    wrapper.vm.handleSave();
    expect(formValidation).toHaveBeenCalled();
    await wrapper.vm.$nextTick()
    expect(actions[UPDATE_TASK]).toHaveBeenCalled();
    await flushPromises();
    expect(console.error).toHaveBeenCalled();
    expect(ElMessage).toHaveBeenCalledWith({
      message: 'Something went wrong while trying to save the changes, please try again',
      type: 'error',
      duration: 5000,
    });
    expect(actions[ROLLBACK_STATE]).toHaveBeenCalled();
    console.error.mockRestore();
  });

  it('it handles error when deleting a task', async () => {
    jest.spyOn(console, 'error').mockImplementation(() => {});
    const [store, { actions }] = mockStore({ delete_task: jest.fn(() => Promise.reject('error'))});
    const wrapper = createWrapper((dataToMount) => {
      dataToMount.global.plugins = [store];
      dataToMount.global.stubs.ElForm = { template: '<form><slot /></form>' };
      return dataToMount;
    });

    wrapper.vm.form = { name: 'Test Save', category: 'Planned', order: 1000 };
    wrapper.vm.handleDelete();
    await wrapper.vm.$nextTick()
    expect(actions[DELETE_TASK]).toHaveBeenCalled();
    await flushPromises();
    expect(console.error).toHaveBeenCalled();
    expect(ElMessage).toHaveBeenCalledWith({
      message: 'Something went wrong while trying to delete the task, please try again',
      type: 'error',
      duration: 5000,
    });
    expect(actions[ROLLBACK_STATE]).toHaveBeenCalled();
    console.error.mockRestore();
  });
});


const createWrapper = (decorator = (n) => n) => {
  const dataToMount = {
    global: {
      directives: {
        loading: () => {},
      },
      stubs: {
        ElDialog: true,
        ElOption: true,
        ElForm: true,
        ElButton: true,
        ElSelect: true,
        ElFormItem: true,
        ElInput: true,
        ElPopconfirm: true,
        ElInputNumber: true,
      },
    },
  };

  decorator(dataToMount);
  return shallowMount(EditTaskForm, dataToMount);
};

function mockStore({ set_task_to_edit, delete_task, update_task, rollback_state, taskToEdit } = {}) {
  const state = {
    tasks: tasks,
    categories: {
      Planned: { name: 'Planned', color: 'red' },
      'In Progress': { name: 'In Progress', color: 'red' },
      Completed: { name: 'Completed', color: 'red' },
    },
    taskToEdit: taskToEdit || {},
  };

  const actions = {
    [SET_TASK_TO_EDIT]: set_task_to_edit || jest.fn(() => Promise.resolve()),
    [DELETE_TASK]: delete_task || jest.fn(() => Promise.resolve()),
    [UPDATE_TASK]: update_task || jest.fn(() => Promise.resolve()),
    [ROLLBACK_STATE]: rollback_state || jest.fn(() => Promise.resolve()),
  }

  const store = createStore({
    state,
    actions,
  });

  return [store, { actions }];
}
