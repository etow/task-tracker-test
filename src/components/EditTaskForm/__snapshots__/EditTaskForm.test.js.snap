// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditTaskForm.vue renders correctly 1`] = `
<div>
  <el-form-stub model="[object Object]" class="edit-task-form" rules="[object Object]" label-width="120px" label-position="left">
    <el-form-item-stub label="Name" prop="name">
      <el-input-stub modelvalue=""></el-input-stub>
    </el-form-item-stub>
    <el-form-item-stub label="Description" prop="description">
      <el-input-stub modelvalue="" type="textarea"></el-input-stub>
    </el-form-item-stub>
    <el-form-item-stub label="Status" prop="category">
      <el-select-stub modelvalue="" style="width: 100%;">
        <el-option-stub value="Planned"></el-option-stub>
        <el-option-stub value="In Progress"></el-option-stub>
        <el-option-stub value="Completed"></el-option-stub>
      </el-select-stub>
    </el-form-item-stub>
    <el-form-item-stub label="Estimate" prop="estimate">
      <el-input-number-stub modelvalue="1" controls-position="right" min="0"></el-input-number-stub>
    </el-form-item-stub>
    <el-form-item-stub style="margin-left: -120px; padding-top: 20px;">
      <div style="display: flex; width: 100%;">
        <el-popconfirm-stub title="Are you sure to delete this task?" hide-icon="true" confirm-button-type="danger"></el-popconfirm-stub>
        <div style="margin-left: auto;">
          <el-button-stub>Cancel</el-button-stub>
          <el-button-stub type="primary">Save</el-button-stub>
        </div>
      </div>
    </el-form-item-stub>
  </el-form-stub>
</div>
`;
